client/state/data-layer/http-data.ts

```diff
@@ -0,0 +1,345 @@
+/**
+ * External dependencies
+ */
+import { Reducer, AnyAction, Dispatch, Action, StoreEnhancerStoreCreator } from 'redux';
+
+/**
+ * Internal dependencies
+ */
+import { HTTP_DATA_REQUEST, HTTP_DATA_TICK } from 'state/action-types';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+
+/**
+ * Types
+ */
+import { Lazy, TimestampMS, TimerHandle } from 'types';
+
+enum DataState {
+	Failure = 'failure',
+	Pending = 'pending',
+	Success = 'success',
+	Uninitialized = 'uninitialized',
+}
+
+interface ResourceData {
+	state: DataState;
+	data: any;
+	error: any;
+	lastUpdated: TimestampMS;
+	pendingSince: TimestampMS | undefined;
+}
+
+type Resource =
+	| ResourceData & {
+			state: DataState.Uninitialized;
+			data: undefined;
+			error: undefined;
+			pendingSince: undefined;
+	  }
+	| ResourceData & { state: DataState.Pending; error: undefined }
+	| ResourceData & { state: DataState.Failure; pendingSince: undefined }
+	| ResourceData & { state: DataState.Success; error: undefined; pendingSince: undefined };
+
+type DataId = string;
+
+export const httpData = new Map< DataId, Resource >();
+export const listeners = new Set< () => void >();
+
+export const empty: Readonly< Resource > = Object.freeze( {
+	state: DataState.Uninitialized,
+	data: undefined,
+	error: undefined,
+	lastUpdated: -Infinity,
+	pendingSince: undefined,
+} );
+
+export const getHttpData = ( id: DataId ) => httpData.get( id ) || empty;
+
+export const subscribe = ( f: () => void ): ( () => void ) => {
```

Comment:
Okay, I'll admit I'm not sure what the second `: ( () => void )` is for :thinking: This is kind of in an unexpected place. Is that the type of our `subscribe` function? :thinking:

Reply:
Yep! That's the return type which is a function that can be called to unsubscribe.